### PR TITLE
Use Gradle instead of Maven

### PR DESCRIPTION
**Pull request by Godin**
# :warning: DO NOT MERGE :construction: WIP :warning:

- [x] pass unit tests
- [x] pass light integration tests (executes Maven on subproject that uses BOM :exclamation:)
- [x] deploy to Repox
- [x] pass all integration tests (requires deployment of `sonar-xoo-plugin`)
- [x] `run-db-unit-tests.sh` executes Maven on `sonar-db-core`, `sonar-db-migration` and `sonar-db-dao` (see `run-db-unit-tests.sh`) and so requires deployment of their dependencies, i.e. `sonar-core`, `sonar-scanner-protocol`, `sonar-process` and `sonar-testing-harness`, but should execute Gradle as discussed with Simon
    - is it normal that `:server:sonar-db-core:test` and `:server:sonar-db-migration:test`
 pass without creation of DB and only `:server:sonar-db-dao:test` fails? => according to Seb this is expected
    - uses `orchestrator-maven-plugin` to create DB :exclamation: 
- [x] get rid of POMs
- [x] get rid of BOM
- [ ] `sonarlint-core` and `sonar-scanner-api` depend on `sonar-markdown`, `sonar-scanner-engine`, `sonar-scanner-protocol`
- [ ] compare artifacts
- [ ] SQ analysis
- [ ] add license check?
- [ ] check IDE integration ❗️ 
    - IntellIJ: needed to manually mark as `Generated Sources Root`
        - `server/sonar-process/build/generated/source/proto/main/java`
        - `sonar-ws/build/generated/source/proto/main/java`
- [ ] `buildVersion` can be different from `project.version` ❗️ 

Can be done later:
- sourcesJar, javadocJar
- `sonar-colorizer` and `sonar-testing-harness` create files in `target` directory
- `runBenchmarks` profile in  `sonar-scanner-engine`
- `scanner-report-viewer` in `sonar-scanner-protocol`
- branding for `sonar-web`
- randomize timezone for unit tests

## Notable differences

- Gradle downloads artifacts into its own cache, not into Maven local repository even if they are downloaded from Maven Central
- system properties in tests are isolated from system properties in Gradle, which is not the case with Maven
- `build/` is a new `target/`

### CODE REVIEW

File: plugins/sonar-xoo-plugin/build.gradle [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+jar {
+  classifier = 'original'
+}
+shadowJar {
+  configurations = [project.configurations.default]
+  // location and name is important for integration tests
+  destinationDir = file('target')
+  classifier = null
+  manifest {
```
Comment (henryju) [inline plugins/sonar-xoo-plugin/build.gradle]: There is a dedicated plugin to replace sonar-maven-packaging-plugin. Not sure about its quality: https://github.com/iwarapter/gradle-sonar-packaging-plugin

File: sonar-scanner-engine/build.gradle [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  testCompile 'junit:junit'
+  testCompile 'org.assertj:assertj-core'
+  testCompile 'net.javacrumbs.json-unit:json-unit-assertj:0.0.15'
+  testCompile 'org.mockito:mockito-core'
+  testCompile 'org.hamcrest:hamcrest-all'
+  testCompile project(':sonar-plugin-api').sourceSets.test.output
+  testCompile 'org.eclipse.jetty:jetty-server'
+  testCompile 'javax.servlet:javax.servlet-api'
+  // TODO how it is used?
```
Comment (henryju) [inline sonar-scanner-engine/build.gradle]: https://github.com/SonarSource/sonarqube/blob/master/sonar-scanner-engine/src/test/java/org/sonar/scanner/mediumtest/fs/FileSystemMediumTest.java#L61

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+allprojects {
+  apply plugin: 'com.jfrog.artifactory'
+
+  group = 'org.sonarsource.sonarqube'
+  version = '6.4' + versionSuffix
+
+  repositories {
+    mavenLocal()
+    mavenCentral()
```
Comment (simonbrandhof) [inline build.gradle]: I suggest to replace `mavenCentral()` by 
```
maven {
  url "https://repox.sonarsource.com/sonarsource"
}
``` 

This superset of Maven Central contains additional SonarSource builds and network access is more appropriate from SonarSource offices.

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#Thu Mar 02 00:59:20 CET 2017
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.4-bin.zip
```
Comment (simonbrandhof) [inline gradle/wrapper/gradle-wrapper.properties]: `gradle-3.4.1-bin.zip` is out!

Comment (Godin) [inline gradle/wrapper/gradle-wrapper.properties]: ✅

File: build.gradle (hunk 3 above)
Comment (Godin) [inline build.gradle]: Done

**Pull request closed without merging**